src/snippets/snippets.ts

```diff
@@ -1,58 +1,73 @@
+ * 
+ * `regex` snippets support string (with the "r" raw option set) or regex triggers, and string or function replacements.
+ * regex replacement functions take in the regex match and return a string.
+ * 
+ * `string` snippets support string triggers (when no "r" raw option set), and string or function replacements.
+ * string replacement functions take in the matched string and return a string.
+ */
+export type SnippetType = Snippet["type"];
+
+interface _Snippet {
+	type: string;
+	options: Options;
 	priority?: number;
 	description?: string;
-	flags?: string;
 }
 
-export class ParsedSnippet {
+interface VisualSnippet extends _Snippet {
+	type: "visual";
 	trigger: string;
-	replacement: string;
-	options: Options;
-	flags: string;
-	priority?: number;
-	description?: string;
-	
-	constructor(raw: RawSnippet) {
-		// normalize regex triggers
-		const resolved: RawSnippet = { ...raw, flags: raw.flags ?? "" };
-		if (raw.trigger instanceof RegExp) {
-			resolved.options = `r${raw.options}`;
-			resolved.trigger = raw.trigger.source;
-			// regex trigger flags and snippet flags get merged 
-			resolved.flags = `${raw.trigger.flags}${resolved.flags}`;
-		}
-
-		// filter out invalid flags
-		const validFlags = [
-			// "d", // doesn't affect the search
-			// "g", // doesn't affect the pattern match and is almost certainly undesired behavior
-			"i",
-			"m",
-			"s",
-			"u",
-			"v",
-			// "y", // almost certainly undesired behavior
-		];
-		resolved.flags = Array.from(new Set(resolved.flags.split("")))
-			.filter(flag => validFlags.includes(flag))
-			.join(""); 
-
-		const parsed = {...resolved, options: Options.fromSource(resolved.options)};
-		Object.assign(this, parsed);
-	}
+	replacement: string | Fn<[selection: string], string | false>;
 }
 
-export interface Environment {
-	openSymbol: string,
-	closeSymbol: string
+interface RegexSnippet extends _Snippet {
+	type: "regex";
+	trigger: string;
+	replacement: string | Fn<[match: RegExpExecArray], string>;
+	flags?: string;
+}
+
+interface StringSnippet extends _Snippet {
+	type: "string";
+	trigger: string;
+	replacement: string | Fn<[match: string], string>;
+	flags?: string;
 }
 
-export const EXCLUSIONS:{[trigger: string]: Environment} = {
-	"([A-Za-z])(\\d)": {openSymbol: "\\pu{", closeSymbol: "}"},
-	"->": {openSymbol: "\\ce{", closeSymbol: "}"}
+/**
+ * serialize some snippet-like value.
+ * specifically, does a pretty-printed JSON.stringify (2-space indent)
+ * that serializes functions to the string "[[Function]]"
+ */
+export function serializeSnippet(snippet: unknown): string {
+	function replacer(k: string, v: unknown) {
+		if (typeof v === "function") { return "[[Function]]"; }
+		if (v instanceof RegExp) { return `[[RegExp]]: /${v.source}/${v.flags}`; }
+		return v;
+	}
+	return JSON.stringify(snippet, replacer, 2);
 }
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+type Fn<Args extends readonly any[], Ret> = (...args: Args) => Ret;
```

Comment:
Not super familiar with TS -- is there a reason for defining this `Fn` type, as opposed to simply writing out the function signatures for each of the `replacement`s?

Reply:
yeah it's not necessary, just there to make function signature types more concise.
would you prefer it be fully written out?

Reply:
yes, thanks